### ISSUE TITLE
Input tokens for the generate function

### ISSUE THREAD

**Issue by jblamare** [Alpha-VLLM/LLaMA2-Accessory#68]
Hello and thank you for your work on this repo!

I had a question about the `generate` function as implemented in the [MetaModel](https://github.com/Alpha-VLLM/LLaMA2-Accessory/blob/main/accessory/model/meta.py). In particular, on [this line](https://github.com/Alpha-VLLM/LLaMA2-Accessory/blob/main/accessory/model/meta.py#L124) where the logits are computed:
```
logits = self.llma.forward_inference(tokens[:, prev_pos:cur_pos], prev_pos, images if prev_pos == 0 else None)
```
Assuming I have a single prompt with 10 tokens (`start_pos = min_prompt_size = 10`), the values for prev_pos/cur_pos will be the following over the first few iterations:
| prev_pos    | cur_pos |
| -------- | ------- |
| 0  | 10   |
| 10 | 11    |
| 11    | 12    |

Meaning that apart from the first token, `tokens[:, prev_pos:cur_pos]` will always be a single token. Doesn't that mean that the model does not have access to the full prompt & history when generating a new token? Shouldn't it be `tokens[:,:cur_pos]`?

I think I'm probably missing something since the generation actually works fine, but I can't see what.

**Comment by ChrisLiu6** [comment]
Hi, a cache mechanism exists in the implementation of every model in `[acccessory/model/LLM](https://github.com/Alpha-VLLM/LLaMA2-Accessory/tree/main/accessory/model/LLM)`. For example, in `llama.py`, you can find the cache [here](https://github.com/Alpha-VLLM/LLaMA2-Accessory/blob/main/accessory/model/LLM/llama.py#L165).

Briefly, in each attention layer, the cache stores the key & values of past tokens, and then when new tokens together with the `start_pos` argument come, they will attend to all cached keys and values before `start_pos`.

**Comment by jblamare** [comment]
Ah that makes more sense now, thank you for your answer!
